### DISCUSSION TITLE
Could sharpen become a part of the image extension?

### DISCUSSION THREAD

**Discussion by PetersOtto** [datenstrom/community#941] [Ask a question] [answered]
Hello!

As I have described [here](https://github.com/datenstrom/yellow/discussions/823#discussion-4628395), I change the image extension for myself so that the images are sharpened. 

Unfortunately, these changes are lost when the image extension is updated.
Therefore my question: Could sharpen become a part of the image extension?

You could activate the function »sharpen« via `system.ini`.

This is what my code looks like:

```
    // Handle initialisation
    public function onLoad($yellow) {
        $this->yellow = $yellow;
       ...
        $this->yellow->system->setDefault("imageSharpenOrNot", "yes");
    }
```
and

```
    // Handle media file changes
    ...
      $sharpenQuery = $this->yellow->system->get("imageSharpenOrNot");
      $image = $this->loadImage($fileName, $type);
      $image = $this->resizeImage($image, $widthInput, $heightInput, $widthOutput, $heightOutput);
      $image = $this->orientImage($image, $orientation);
       if ($type=="jpg" && $sharpenQuery=="yes") {
         $image = $this->sharpenImage($image);
        }
...
```
and
```
    // sharpen image
    public function sharpenImage($image) {
        $sharpen = array([0, -2, 0], [-2, 11, -2], [0, -2, 0]);
        imageconvolution($image, $sharpen, 3, 0);
        return $image;
    }
```

Here are two example images:
![not-sharpen](https://github.com/datenstrom/yellow/assets/52236754/a8bd78b3-e039-4652-9393-4eeb0fc64c1f)
![sharpen](https://github.com/datenstrom/yellow/assets/52236754/c44440ee-d284-42ab-aaf2-07899060b15b)

What do you think?

**Comment by GiovanniSalmeri** [comment]
What I am going to say is not directely related to @PetersOtto's proposal, but the purpose is similar. Some time ago I read an article about HDPI displays (unfortunately I cannot find it anymore) which suggested a very simple trick in order to have better JPG images without increasing the filesize: double the dimensions of the image (and let the browser resize it) and increase the compression. This is a quick example: https://mondodomani.org/img-test/img-test.html

Actually the second image (which has a slightly smaller filesize!) seems to me in a HDPI display clearly better than the first. This would imply a simple modification of the logic of image.php. Perhaps this is another possibility to evaluate?

Edit: this is not the article I read, but the same trick is suggested: https://web.dev/articles/easy-high-dpi-images

**Reply by PetersOtto** [reply]
Hi Giovanni, I haven't heard of this trick before. Interesting! I will try the trick. Thank you!
However, I don't think this is about sharpening images. It's about a better/smarter display of images. 
Perhaps both should be combined?   &#128512;

Sharpening becomes important when high resolution images are greatly reduced in resolution.

If you decrease images, the images going unsharp because some pixel gone during this process. If needed, it is possible to sharp the image after decreasing the image. Usually you use the »unsharp mask« filter. (Photoshop, Affinity Photo ... )

**Comment by markseuffert** [comment]
When I ask myself, do I need this, do I want this? My answer would probably be no. I would lean slightly towards putting image editing functionality in its own extension. For example in a new imagefilter extension for uploaded images. Maybe with different filters for sharpening and softening photos? Maybe with automatic adjustments of color and contrast? Maybe with selfie enhancements... there are many possibilities.

**Comment by PetersOtto** [comment]
Thanks for your feedback!
You mean an extension that creates a copy of the original image and modifies it, after it has been uploaded?

What I currently miss about the image extension is the flexibility. It would be great to have the following points. (In my opinion, it is just a discussion &#128522;)

* loading="lazy" --> on/off
* title="alt text" --> on/off
* filter --> on/off and implement new, own filter (sharpen, blur, etc)

Maybe it would possible to add an interface to the image extension? (... `yellow->extension->isExisting("image-filter-plugin")`)
Only if the extension »image-filter-plugin« is present, a filter will be used?

**Reply by markseuffert** [reply, marked as answer]
It depends on what you want to do. The [image extension](https://github.com/annaesvensson/yellow-image) was created to display images with Markdown. I personally think we should not support everything that's technical possible. Product design is partially about good technology, partially about finding what's good enough. Whenever a small solution handles 80% and a big solution handles 100%, but has the cost of additional complexity, then it's worth to reconsider the additional effort. Sometimes less is more. Or as Dieter Rams describes good design, less but better. [[1]](https://www.interaction-design.org/literature/article/dieter-rams-10-timeless-commandments-for-good-design) [[2]](https://www.youtube.com/watch?v=-mlAdcQZZAM) 

The good thing is that there is a [modular extension system](https://datenstrom.se/yellow/extensions/) and we don't have to put everything into one image extension. My suggestion would be to make your own "imageextended" extension or split the functionality into multiple extensions, for example "image" and "imagefilter". Then people can choose what they want to install. You can also experiment with different ideas and features. If something becomes popular you can [publish it](https://github.com/annaesvensson/yellow-publish).

Hope this helps and see also datenstrom/community#892.

**Reply by PetersOtto** [reply]
Accepted &#128516;

**Reply by markseuffert** [reply]
Thanks 😄
